lib/sensors/src/optical_flow.rs

```diff
@@ -0,0 +1,365 @@
+
+            let dr = cursor.next().unwrap();
+            let _obs = cursor.next().unwrap();
+            let x = i16::from_le_bytes([*cursor.next().unwrap(), *cursor.next().unwrap()]);
+            let y = i16::from_le_bytes([*cursor.next().unwrap(), *cursor.next().unwrap()]);
+            let quality = *cursor.next().unwrap();
+            let _raw_sum = cursor.next();
+            let _raw_max = cursor.next();
+            let _raw_min = cursor.next();
+            let shutter_upper = *cursor.next().unwrap();
+            let _shutter_lower = cursor.next();
+
+            // Validate the data
+            if (dr & 0b1000_0000) != 0 && !(quality < 0x19 && shutter_upper == 0x1F) {
+                return Ok(Motion { x, y });
+            }
+
+            // Wait before retrying
+            Timer::after(RETRY_DURATION).await;
+        }
+
+        Err("Timed out waiting for motion data")
+    }
+
+    /// Set orientation of PMW3901 in increments of 90 degrees.
+    pub async fn set_rotation(&mut self, orientation: Orientation) -> Result<(), OpticalFlowError> {
+        match orientation {
+            Orientation::Degrees0 => self.set_orientation(true, true, true).await,
+            Orientation::Degrees90 => self.set_orientation(false, true, false).await,
+            Orientation::Degrees180 => self.set_orientation(false, false, true).await,
+            Orientation::Degrees270 => self.set_orientation(true, false, false).await,
+        }
+    }
+
+    /// Sets the orientation of the PMW3901 manually.
+    /// Swapping is performed before flipping.
+    pub async fn set_orientation(
+        &mut self,
+        invert_x: bool,
+        invert_y: bool,
+        swap_xy: bool,
+    ) -> Result<(), OpticalFlowError> {
+        let mut value = 0;
+        if swap_xy {
+            value |= 0b1000_0000;
+        }
+        if invert_x {
+            value |= 0b0100_0000;
+        }
+        if invert_y {
+            value |= 0b0010_0000;
+        }
+        self.write(REG_ORIENTATION, value)
+    }
+
+    /// Writes a single byte to a register
+    fn write(&mut self, register: u8, data: u8) -> Result<(), OpticalFlowError> {
+        self.cs.set_low();
+        let result = match self.spi.transfer_in_place(
+            // OR 0x80 to the register
+            &mut [register | 0x80, data],
+        ) {
+            Ok(_) => Ok(()),
+            Err(e) => Err(OpticalFlowError::SpiError(e)),
+        };
+        self.cs.set_high();
+        result
+    }
+
+    /// Read a single byte from a register
+    fn read(&mut self, register: u8) -> Result<u8, OpticalFlowError> {
+        let data = &mut [register, 0];
+        self.cs.set_low();
+        self.spi.transfer_in_place(data).unwrap();
+        self.cs.set_high();
+        Ok(data[1])
+    }
+
+    /// Perform a bulk write of data to the sensor
+    async fn bulk_write(&mut self, write_data: &[(u8, u8)]) -> Result<(), OpticalFlowError> {
```

Comment:
interesting that this seems to set and release cs for each write, that was what i thought went wrong with the c++ last year